File: cpp/Scanner.cpp
#include <unordered_map>
#include "Lox.h"
#include "Scanner.h"
#include "Token.h"
#include "TokenType.h"
#include "Object.h"

// Bruh.
// static const std::unordered_map<std::string, TokenType> keywords;
// static const std::unordered_map<std::string, TokenType> keywords = {
// std::unordered_map<std::string, TokenType> keywords = {
//     { "and",    AND },
//     { "class",  CLASS },
//     { "else",   ELSE },
//     { "false",  FALSE },
//     { "for",    FOR },
//     { "fun",    FUN },
//     { "if",     IF },
//     { "nil",    NIL },
//     { "or",     OR },
//     { "print",  PRINT },
//     { "return", RETURN },
//     { "super",  SUPER },
//     { "this",   THIS },
//     { "true",   TRUE },
//     { "var",    VAR },
//     { "while",  WHILE }
// };

Scanner::Scanner(std::string source) {
    this->source = source;
    // I think tokens is already initialized
}

std::list<Token> Scanner::scanTokens() {
    while (not isAtEnd()) {
        start = current;
        scanToken();
    }

    tokens.push_back(Token(ENDOF, "", Object(), line));
    return tokens;
}

void Scanner::scanToken() {
    char c = advance();
    switch (c) {
        case '(': addToken(LEFT_PAREN); break;
        case ')': addToken(RIGHT_PAREN); break;
        case '{': addToken(LEFT_BRACE); break;
        case '}': addToken(RIGHT_BRACE); break;
        case ',': addToken(COMMA); break;
        case '.': addToken(DOT); break;
        case '-': addToken(MINUS); break;
        case '+': addToken(PLUS); break;
        case ';': addToken(SEMICOLON); break;
        case '*': addToken(STAR); break;
        // Two spaces? Why not.
        case '!':
          addToken(match('=') ? BANG_EQUAL : BANG);
          break;
        case '=':
          addToken(match('=') ? EQUAL_EQUAL : EQUAL);
          break;
        case '<':
          addToken(match('=') ? LESS_EQUAL : LESS);
          break;
        case '>':
          addToken(match('=') ? GREATER_EQUAL : GREATER);
          break;
        case '/':
          if (match('/')) {
              // This is a comment!
              while (peek() != '\n' and not isAtEnd()) { advance(); }
          } else {
              addToken(SLASH);
          }
          break;
        case ' ':
        case '\r':
        case '\t':
          // Ignore whitespace.
          break;
        case '\n':
          line++;
          break;
        case '"': string(); break;
        default:
          if (isDigit(c)) {
            number();
          } else if (isAlpha(c)) {
            identifier();
          } else {
              Lox::error(line, "Unexpected character.");
          }
          break;
    }
}

void Scanner::identifier() {
    while (isAlphaNumeric(peek())) { advance(); }

    std::string text = source.substr(start, current);
    TokenType type;
    // try {
    //     type = keywords.at(text);
    // } catch (std::out_of_range e) {
    //     type = IDENTIFIER;
    // }
    type = IDENTIFIER;
    addToken(type);
}

void Scanner::number() {
    while (isDigit(peek())) { advance(); }

    // Look for a fractional part.
    if (peek() == '.' and isDigit(peekNext())) {
        // Consume the "."
        advance();

        while (isDigit(peek())) { advance(); }
    }

    addToken(NUMBER, std::stod(source.substr(start, current)));
}

void Scanner::string() {
    while (peek() != '"' and not isAtEnd()) {
        if (peek() == '\n') line++;
        advance();
    }

    if (isAtEnd()) {
        Lox::error(line, "Unterminated string.");
        return;
    }

    // The closing ".
    advance();

    // Trim the surrounding quotes.
    std::string value = source.substr(start + 1, current - 1);
    addToken(STRING, value);
}

bool Scanner::match(char expected) {
    if (isAtEnd()) return false;
    if (source[current] != expected) return false;

    current++;
    return true;
}

char Scanner::peek() {
    if (isAtEnd()) return '\0';
    return source[current];
}

char Scanner::peekNext() {
    if (current + 1 >= (int)source.length()) return '\0';
    return source[current + 1];
}

bool Scanner::isAlpha(char c) {
    return (c >= 'a' and c <= 'z') or
           (c >= 'A' and c <= 'Z') or
            c == '_';
}

bool Scanner::isDigit(char c) {
    return c >= '0' and c <= '9';
}

bool Scanner::isAlphaNumeric(char c) {
    return isAlpha(c) or isDigit(c);
}

bool Scanner::isAtEnd() {
    return current >= (int)source.length();
}

char Scanner::advance() {
    return source[current++];
}

void Scanner::addToken(TokenType type) {
    addToken(type, Object());
}

void Scanner::addToken(TokenType type, Object literal) {
    std::string text = source.substr(start, current);
    tokens.push_back(Token(type, text, literal, line));
}

File: ruby/scanner.rb
require "./lox"
require "./token"

class Scanner
  def initialize( source )
    @source = source
    @tokens = []
    @start = 0
    @current = 0
    @line = 1

    @keywords = { 
      'and' => :and,
      'class' => :class,
      'else' => :else,
      'false' => :false,
      'for' => :for,
      'fun' => :fun,
      'if' => :if,
      'nil' => :nil,
      'or' => :or,
      'print' => :print,
      'return' => :return,
      'super' => :super,
      'this' => :this,
      'true' => :true,
      'var' => :var,
      'while' => :while
    }
  end

  def scanTokens
    until isAtEnd
      @start = @current
      scanToken
    end

    @tokens << Token::new(:eof, "", nil, @line)
    return @tokens
  end

  def isAtEnd
    @current >= @source.length
  end

  def scanToken
    c = advance
    # switch statement
    case c
    when '('
      addToken(:left_paren)
    when ')'
      addToken(:right_paren)
    when '{' 
      addToken(:right_brace)
    when '}' 
      addToken(:right_brace)
    when ',' 
      addToken(:comma)
    when '.' 
      addToken(:dot)
    when '+' 
      addToken(:plus)
    when '-' 
      addToken(:minus)
    when '*' 
      addToken(:star)
    when ';' 
      addToken(:semicolon)
    when '!'
      addToken(:bang_equal ? match('=') : :bang)
    when '='
      addToken(:equal_equal ? match('=') : :equal)
    when '>'
      addToken(:greater_equal ? match('=') : :greater)
    when '<'
      addToken(:less_equal ? match('=') : :less)
    when '/'
      if match('/')
        # A comment goes until the end of the line.
        while peek != "\n" && !isAtEnd
          advance
        end
      else
        addToken(:slash)
      end
    when " "
    when "\r"
    when "\t"
    when "\n"
      @line += 1
    when '"'
      string
    else
      if isDigit(c)
        number
      elsif isAlpha(c)
        identifier
      else
        Lox::error(@line, "Unexpected character.")
      end
    end
  end

  def advance
    curr = @source[@current]
    @current += 1
    return curr
  end

  def addToken( *args )
    case args.size
    when 1
      # def addToken( type )
      #   addToken(type, nil)
      # end
      addToken(args[0], nil)
    when 2
      # def addToken( type, literal )
      #   text = @source[@start, @current]
      #   @tokens << Token.new(type, text, literal, @line)
      # end
      text = @source[@start..@current - 1]
      @tokens << Token.new(args[0], text, args[1], @line)
    end
  end

  def match( expected )
    if isAtEnd
      return false
    end

    if @source[@current] != expected
      return false
    end

    @current += 1
    return true
  end

  def peek
    if isAtEnd
      return nil
    end
      return @source[@current]
  end

  def peekNext
    if @current + 1 >= @source.length
      return nil
    end
    return @source[@current + 1]
  end

  def string
    while peek != '"' && !isAtEnd
      if peek == "\n"
        @line += 1
      end
      advance
    end
    
    if isAtEnd
      Lox::error(@line, "Unterminated string.")
      return
    end

    # The closing ".
    advance

    # Trim the surrounding quotes.
    value = @source[@start + 1..@current - 2]
    addToken(:string, value)
  end

  def isDigit( c )
    return ('0'..'9') === c
  end

  def isAlpha( c )
    return ('a'..'z') === c || ('A'..'Z') === c || c == "_"
  end

  def isAlphaNumeric( c )
    return isDigit(c) || isAlpha(c)
  end

  def identifier
    while isAlphaNumeric peek
      advance
    end
    text = @source[@start..@current - 1]
    type = @keywords[text]
    unless type
      type = :identifier
    end
    addToken(type)
  end

  def number
    while isDigit( peek )
      advance
    end

    # Look for a fractional part.
    if peek == '.' && isDigit( peekNext )
      # Consume the "."
      advance
      while isDigit( peek )
        advance
      end
    end

    addToken(:number, @source[@start..@current - 1].to_f)
  end

end
